src/resources/nfts/index.ts

```diff
@@ -0,0 +1,170 @@
+  let finished = false;
+  let cursor: string | undefined;
+  let freshNFTs: UniqueAsset[] = [];
+
+  while (!finished) {
+    // eslint-disable-next-line no-await-in-loop
+    const [simplehashResponse, polygonAllowlist] = await Promise.all([
+      fetchSimplehashNFTs(address, cursor),
+      fetchPolygonAllowlist(),
+    ]);
+
+    const { nfts: simplehashNFTs, nextCursor } = simplehashResponse;
+
+    const newNFTs = simplehashNFTs
+      .filter(nft => {
+        if (nft.chain === SimplehashChain.Polygon) {
+          return polygonAllowlist.includes(nft.contract_address);
+        }
+        return true;
+      })
+      .map(simplehashNFTToUniqueAsset);
+
+    freshNFTs = [...newNFTs, ...freshNFTs];
+
+    if (nextCursor && freshNFTs.length < NFTS_LIMIT) {
+      cursor = nextCursor;
+    } else {
+      // eslint-disable-next-line require-atomic-updates
+      finished = true;
+    }
+
+    const currentNFTs =
+      queryClient.getQueryData<UniqueAsset[]>(nftsQueryKey({ address })) ?? [];
+
+    // iteratively update query data with new NFTs until the limit is hit
+    if (currentNFTs.length < NFTS_LIMIT) {
+      queryClient.setQueryData<UniqueAsset[]>(
+        nftsQueryKey({ address }),
+        cachedNFTs => uniqBy([...newNFTs, ...(cachedNFTs ?? [])], 'uniqueId')
+      );
+    }
+  }
+
+  // once we successfully fetch all NFTs, replace all cached NFTs with fresh ones
+  queryClient.setQueryData<UniqueAsset[]>(
+    nftsQueryKey({ address }),
+    () => freshNFTs
+  );
+
+  return freshNFTs;
+}
+
+export function useNFTs(): NFT[] {
+  // normal react query where we get new NFT formatted data
+  return [];
+}
+
+export function useLegacyNFTs(
+  address: string
+): { nfts: UniqueAsset[]; isLoading: boolean } {
+  const queryClient = useQueryClient();
+  const mounted = useIsMounted();
+
+  const [cursor, setCursor] = useState<string>();
+  const [isFinished, finish] = useReducer(() => true, false);
+  const [freshNFTs, setFreshNFTs] = useState<UniqueAsset[]>([]);
+
+  const queryKey = nftsQueryKey({ address });
+
+  // listen for query udpates
+  const query = useQuery<UniqueAsset[]>(queryKey, async () => [], {
+    enabled: false,
+    staleTime: Infinity,
+  });
+
+  const nfts = query.data ?? [];
+
+  useEffect(() => {
+    // stream in NFTs one simplehash response page at a time
+    const fetchNFTs = async () => {
+      const [simplehashResponse, polygonAllowlist] = await Promise.all([
+        fetchSimplehashNFTs(address, cursor),
+        fetchPolygonAllowlist(),
+      ]);
+
+      const { nfts: simplehashNFTs, nextCursor } = simplehashResponse;
+
+      const newNFTs = simplehashNFTs
+        .filter(nft => {
+          if (nft.chain === SimplehashChain.Polygon) {
+            return polygonAllowlist.includes(nft.contract_address);
+          }
+          return true;
+        })
+        .map(simplehashNFTToUniqueAsset);
+
+      const updatedFreshNFTs = [...newNFTs, ...freshNFTs];
+      setFreshNFTs(updatedFreshNFTs);
```

Comment:
Do we need to use `uniqBy` here too? Also, should we append or prepend the new NFTs? Rn it's prepending, which I feel like might be more render-intensive since it has to re-layout the whole list every time. Unsure.

Reply:
i don't think we need to use `uniqBy` here, since `freshNFTs` only contains brand new data and doesn't mix with the cache at all. there's a very small chance that simplehash's response updates in between page fetches, but i think that's unrealistic enough that it's not worth the performance hit

Reply:
appending is more efficient, and i found that using `concat` is slightly faster than using the spread operator so I'll make those changes.